File: ruby/lightsabers/refactor.rb
'polymorphism'

class Lightsaber < ActiveRecord::Base
  def set_color(warrior)
    self.color = pick_color(warrior.class)
  end

  WARRIOR_COLORS = {
    Jedi =>    ['green', 'blue', 'purple'],
    Sith =>    ['red'],
    Kaleesh => ['green', 'blue']
  }

  def pick_color(klass)
    WARRIOR_COLORS[klass].sample
  end
end

File: csharp/lightsabers/refactor.cs
"polymorphism.cs"

class Lightsaber
{
  public void AfterAdd(object klass)
  {
    this.color = PickColor(klass.getType());
  }

  private const Dictionary<string, string[]> warriors = return new Dictionary<string, string[]>()
  {
    { "Jedi", { "green", "blue", "purple" } },
    { "Sith", { "red" } },
    { "Kaleesh", { "green", "blue" } }
  }

  private string PickColor(string klass_name)
  {
    string[] colors = warriors[klass_name];
    return colors[random.Next(0, colors.length)]
  }
}
